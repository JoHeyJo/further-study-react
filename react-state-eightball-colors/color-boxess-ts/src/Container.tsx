import React, { useState } from 'react';
import Box from './Box';
import { randomNum, chooseRandom } from './utils';
import './Container.css';


const defaultColors = [
  "Aqua", "Black", "BlanchedAlmond", "Blue", "Chocolate", "DodgerBlue",
  "Lavender", "LawnGreen", "Peru", "Plum", "SpringGreen", "SteelBlue", "Tan",
  "Teal", "Thistle", "Tomato", "Turquoise", "Violet", "Yellow", "YellowGreen",
];

/** holds all box components
 * 
 * Props:
 * - colors
 * - numBoxes: default number of boxes
 * 
 * State:
 * - boxes: []
 */
function Container({ numBoxes = 16 }) {
  const [colors, setColors] = useState<string[]>(createInitialBoxes)
  const [index, setIndex] = useState<number | null>(null)

  /** Return array of random colors */
  function createInitialBoxes() {
    return Array.from(
      { length: numBoxes },
      () => chooseRandom(defaultColors)
    )
  }

  /** on button click: changes color of selected box */
  function changeColor(): void {
    const newColor = chooseRandom(colors);
    const idx = randomNum(colors.length);
    setIndex(idx);
    setColors(c => {
      let copy = [...c] // don't forget to create copy of state before mutating
      copy[idx] = newColor
      return copy
    })
  }

  const boxes = colors.map((color, i) => <Box key={i} color={color} isChanged={i === index} />)

  return (
    <div>
      <div className='Container-box'>
        {boxes}
      </div>
      <button onClick={changeColor}>Change</button>
    </div>

  )
}

export default Container;