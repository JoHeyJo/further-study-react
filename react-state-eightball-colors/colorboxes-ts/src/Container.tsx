import React, { useState } from 'react';
import Box from "./Box";
import './container.css'

function randomNum(num: number) {
  return Math.floor(Math.random() * num)
}

function choice(colors: string[]) {
  return colors[randomNum(colors.length)]
}

interface IContainer {
  colors: string[];
  numBoxes: number;
}

/** Container holding all the Box components
 * 
 * 
 * Props:
 * - colors: array or random colors
 * - numBoxes: num of boxes to manage
 * 
 * state:
 * - boxes: [color,color,color...]
 * 
 * 
 */

function Container({ colors, numBoxes }: IContainer) {
  const [boxes, setBoxes] = useState(getInitialRandomColors);

  /** populates initial state with 16 boxes */
  function getInitialRandomColors() {
    return Array.from(
      { length: numBoxes },
      () => choice(colors));
  }

  /** choose random box and change color */
  function randomBox() {
    setBoxes(boxes => {
      let idx = randomNum(numBoxes);
      let boxCopy = [...boxes];
      boxCopy[idx] = choice(colors);
      return boxCopy;
    })
  }

  const boxComponent = boxes.map((color, i) => <Box key={i} color={color} />)

  return (
    <div>
      <div className='Container-box'
        onClick={randomBox}>
        {boxComponent}
      </div>
    </div>
  )
}

export default Container;
