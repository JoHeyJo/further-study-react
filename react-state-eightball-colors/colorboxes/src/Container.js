import React, { useState } from "react";
import Box from './Box.js';
import './container.css';
import { randInt, choice } from "./utils.js";


// At the bottom of all of the boxes should be a button labeled “Change”.

// Initially, each box should have a background color chosen from a random list of colors.

// When you click the button:

// it should select a random box
// it should change the background color of that random box to a new color from the possible colors list.
// Before building anything, think about the structure of your React app.This entire thing shouldn’t be one giant component.

/** Container holds all boxes and interactions
 * 
 * 
 * Props:
 * - colors: array of random colors
 * - numBoxes: # of boxes to manage
 * 
 * State:
 * - boxes: array of box colors
 * 
 */
function Container({ colors = defaultColors, numBoxes = 16 }) {
  const [boxes, setBoxes] = useState(getInitialRandomColors);


  /** Return array of random colors */
  function getInitialRandomColors(){
    return Array.from(
      { length: numBoxes },
      () => choice(colors));
  }

 /** On button click: pick random box, change to random color */
 function handleClick(evt){
  setBoxes(boxes => {
    let idx = randInt(numBoxes);
    let boxCopy = [...boxes];
    boxCopy[idx] = choice(colors);
    return boxCopy;
  }); 
 }

 const boxComponents = boxes.map((color, i) => <Box key={i} color={color}/>)

  return (
    <div className="Container-interface">
      <div className="Container-box">
        {boxComponents}
      </div>
      <button onClick={handleClick}>Change</button>
    </div>
  )
}

export default Container;


const defaultColors = ['#FF6633', '#FFB399', '#FF33FF', '#FFFF99', '#00B3E6',
  '#E6B333', '#3366E6', '#999966', '#99FF99', '#B34D4D',
  '#80B300', '#809900', '#E6B3B3', '#6680B3', '#66991A',
  '#FF99E6', '#CCFF1A', '#FF1A66', '#E6331A', '#33FFCC',
  '#66994D', '#B366CC', '#4D8000', '#B33300', '#CC80CC',
  '#66664D', '#991AFF', '#E666FF', '#4DB3FF', '#1AB399',
  '#E666B3', '#33991A', '#CC9999', '#B3B31A', '#00E680',
  '#4D8066', '#809980', '#E6FF80', '#1AFF33', '#999933',
  '#FF3380', '#CCCC00', '#66E64D', '#4D80CC', '#9900B3',
  '#E64D66', '#4DB380', '#FF4D4D', '#99E6E6', '#6666FF'];
