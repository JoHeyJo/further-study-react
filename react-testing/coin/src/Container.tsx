//modules
import React , { useState } from 'react';
//imports
import Coin from './Coin';
import heads from './head.jpeg';
import tails from './tail.jpeg';
import { ICount, ICoin } from './interfaces'

const random = () => Math.floor(Math.random() * 10);

const defaultCount: ICount = {
  heads: 0,
  tails: 0
}

const head: ICoin = {
  side: heads,
  alt: 'heads'
}

const tail: ICoin = {
  side: tails,
  alt: 'tails'
}

function Container(){
  const [coin, setCoin] = useState<ICoin>({side:'',alt:''});
  const [count, setCount] = useState<ICount>(defaultCount)

  /** chooses a heads or tails on 50/50 chance */
  function flipCoin(){
    let side = random() > 4 ? head : tail;
    setCoin(side);
    updateCount(side);
  }

  /** keep track of which side is flipped, updates state object */
  function updateCount(coin:ICoin){
    if(coin.alt === 'heads'){
      setCount(c => {
        c.heads++
        return c
      })
    } else {
      setCount(c => {
        c.tails++
        return c
      })
    }
  }

  return (
    <div>
      <Coin side={coin.side} alt={coin.alt} />
      <p>tails: {count.tails} heads {count.heads}</p>
      <button onClick={flipCoin}>Flip me</button>
    </div>
  )
};

export default Container;