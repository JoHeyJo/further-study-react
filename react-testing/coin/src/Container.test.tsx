// import { expect, jest, test } from '@jest/globals';
import React from "react";
import { fireEvent, getByRole, render } from "@testing-library/react";
import Container from './Container';


beforeEach(function () {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

it("renders without crashing", function (){
  render(<Container />);
});

it("match snapshot", function () {
  const { container } = render(<Container />);
expect(container).toMatchSnapshot();
})

it("show a button 'flip me'", function (){
  const {container } = render(<Container />);
  const button = getByRole(container,'button');
  expect(button).toBeInTheDocument();
})

it("when the button is clicked it shows either heads or tails and tracks it", function (){
  const { container, debug } = render(<Container />);

  const flip = getByRole(container, 'button')
  expect(container.querySelector("img[src=''")).toBeEmptyDOMElement();
  debug()
  //one click for heads
  fireEvent.click(flip)
  expect(container.querySelector("img[src='tail.jpeg']")).toBeInTheDocument();
  expect(container.querySelector("img[src='head.jpeg']")).not.toBeInTheDocument();
  expect(container.querySelector('p')).toHaveTextContent('tails: 1 heads 0');
  //one more click for tails
  fireEvent.click(flip)
  expect(container.querySelector("img[src='head.jpeg']")).toBeInTheDocument();
  expect(container.querySelector("img[src='tail.jpeg']")).not.toBeInTheDocument();
  expect(container.querySelector('p')).toHaveTextContent('tails: 1 heads 1');

  debug()


})

afterEach(function () {
  (Math.random as jest.Mock).mockRestore();
});

