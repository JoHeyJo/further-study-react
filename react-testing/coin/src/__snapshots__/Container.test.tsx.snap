// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div>
  <div>
    <div>
      <img
        alt=""
        src=""
        style="width: 300px;"
      />
    </div>
    <p>
      tails: 
      0
       heads 
      0
    </p>
    <button>
      Flip me
    </button>
  </div>
</div>
`;
